pkg/api/handlers.go

```diff
@@ -261,6 +261,8 @@ func NewQuerierHandler(
 		nil,
 	)
 
+	api.ClearCodecs()
+	api.InstallCodec(mimirJSONCodec{})
 	api.InstallCodec(protobufCodec{})
```

Comment:
```suggestion
	api.ClearCodecs()
	api.InstallCodec(protobufCodec{})
```

Wouldn't this achieve the same? See my overall comment.

Reply:
We could do this in the future, but for now, we need to continue supporting JSON responses in queriers for two scenarios:
* For clusters where query-frontends do not have the protobuf format enabled
* To support easy upgrades between versions - if we remove support for JSON responses in queriers, then we force users to upgrade their query-frontends and enable the protobuf format first, then upgrade queriers separately. If we support both formats, then both components can be rolled out simultaneously, and queriers that support protobuf will only send it to query-frontends that report they support it (via the `Accept` header).

Reply:
My concern is that we're adding a linear algorithm, what is the impact on the performance and resource utilization ? 

I think another way to solve this is to introduce the protobuf encoder, but do not enable native histogram ingestion for anybody, then in the next upgrade turn off the JSON encoder. After the second upgrade you can enable histograms.

Or even better, we could have the MimirJSON encoder be able to encode histograms in the internal format and done.

Reply:
Yep, I agree, I think I've been overzealous here :) We can leave the JSON encoder as it is, and the worst case scenario is that either histograms results disappear (if the query-frontend is too old) or the query-frontend rejects the response containing explicit histograms [here](https://github.com/grafana/mimir/blob/4b43206a721088cdcbdeb1fc943d5797f2b13d45/pkg/frontend/querymiddleware/model_extra.go#L373) (if the query-frontend has been upgraded).